prov/util/src/rxm_av.c

```diff
@@ -205,36 +204,57 @@ static int rxm_av_remove(struct fid_av *av_fid, fi_addr_t *fi_addr,
 			 size_t count, uint64_t flags)
 {
 	struct util_av_entry *av_entry;
+	struct util_peer_addr **peer;
+	struct util_ep *util_ep;
+	struct dlist_entry *item;
 	struct rxm_av *av;
 	ssize_t i;
 
-	av = container_of(av_fid, struct rxm_av, util_av.av_fid);
 	if (flags)
 		return -FI_EINVAL;
 
+	av = container_of(av_fid, struct rxm_av, util_av.av_fid);
+
 	/*
 	 * It's more efficient to remove addresses from high to low index.
 	 * We assume that addresses are removed in the same order that they were
 	 * added -- i.e. fi_addr passed in here was also passed into insert.
 	 * Thus, we walk through the array backwards.
 	 */
-	ofi_mutex_lock(&av->util_av.lock);
 	for (i = count - 1; i >= 0; i--) {
 		FI_INFO(av->util_av.prov, FI_LOG_AV,
 			"fi_addr %" PRIu64 "\n", fi_addr[i]);
+		ofi_mutex_lock(&av->util_av.lock);
 		av_entry = ofi_bufpool_get_ibuf(av->util_av.av_entry_pool,
 						fi_addr[i]);
+		ofi_mutex_unlock(&av->util_av.lock);
 		if (!av_entry)
 			continue;
 
+		if (av->util_av.remove_handler) {
+			ofi_mutex_lock(&av->util_av.lock);
+			peer = ofi_av_addr_context(&av->util_av, fi_addr[i]);
+			(*peer)->refcnt++; /* prevent peer from being freed */
+			ofi_mutex_unlock(&av->util_av.lock);
+			ofi_mutex_lock(&av->util_av.ep_list_lock);
+			dlist_foreach(&av->util_av.ep_list, item) {
+				util_ep = container_of(item, struct util_ep,
+						       av_entry);
+				av->util_av.remove_handler(util_ep, *peer);
+			}
+			ofi_mutex_unlock(&av->util_av.ep_list_lock);
+			util_put_peer(*peer);
+		}
+
 		if (!ofi_atomic_dec32(&av_entry->use_cnt)) {
+			ofi_mutex_lock(&av->util_av.lock);
 			rxm_put_peer_addr(av, fi_addr[i]);
 			HASH_DELETE(hh, av->util_av.hash, av_entry);
 			ofi_ibuf_free(av_entry);
+			ofi_mutex_unlock(&av->util_av.lock);
```

Comment:
The above flow is inefficient and *looks* racy, even if it's not.  The remove_handler is NULL by default.  The original locking structure ends up looking better, plus we should add a comment regarding why we're dropping the lock in the middle of this handling.  Example:
```
lock();
for (i = count - 1; ...) {
    av_entry = ofi_bufpoll_get_ibuf();
    if (!av_entry)
        continue;

    if (remove_handler) {
        /* The remove_handler may call back into the AV to remove the provider's
         * reference on the peer address.  We need to drop the lock on the AV in case
         * the handler tries to acquire it.  Before dropping the lock, we need to increment
         * the reference count on the peer, so that it's still valid to pass into the handler
         * and isn't freed by another thread after we drop the AV lock.
         */
        peer = ofi_av_addr_context();
        peer->refcnt++;
        unlock();
        remove_handler(...);
        util_put_peer(*peer);
        lock()
    }

    if (1ofi_atomic_dec3(...) {
        ....
    }
}
unlock();
```
I'm missing the ep_list_lock and other details, but hopefully that makes sense.

Reply:
It comes down to what util_av.lock is protecting.  To put the lock around the for-loop and then unlock/relock in the middle of it is guaranteed racy, so not doing that.
If you want the lock around the for-loop, that leave one alternative, to have a version of rxm_av_free_conn that does not acquire util_av.lock (since it is already the for-loop).

Reply:
The for loop is acting over parameters passed in through the app.  There's no lock needed for that.  The AV lock is protecting access to the AV pools and data structures.

Reply:
added equivalent functions that do not acquire util_av.lock for use in the for-loop and also in the remove_handler.